### PR TITLE
Add cooldown interval logic for the profiling functional

### PR DESCRIPTION
**Pull request by Icemist**
**Motivation:**

Sometimes running the script several times we get a strong scatter in the results.
```
from tvm.contrib.debugger import debug_runtime as graph_executor
...
m = graph_executor.create(graph, lib, ctx)
...
time_f = m.module.time_evaluator("run", dev, number=1000, repeat=1)
mean = time_f().mean
print(mean)
```

First try:

> 0.0846912 secs/iteration

Second try:
> 0.105601 secs/iteration

In general, the devices has an acceptable temperature range. If the system heats up to the upper limit for some time, it can force cooling by available ways. For example, by lowering the frequency.
In the following example the temperature reaches 90 degrees and after some time the frequency drops and after cooling is restored.
```
from tvm.contrib.debugger import debug_runtime as graph_executor
...
m = graph_executor.create(graph, lib, ctx)
...
time_f = m.module.time_evaluator("run", dev, number=1)
one = time_f().results[0]
nums = round(1/one)
time_f = m.module.time_evaluator("run", dev, number=1, repeat=nums*200)
t =  time_f()
mean = t.mean
print(mean)
0.105511 secs/iteration

print(t.results)
[0.083143681, ... , 0.147928064, ... , 0.147806464]
```
**_the results of device profiling_** 
![66_over](https://user-images.githubusercontent.com/14822750/171618129-a0dfd90e-dea3-4e50-8c12-1b847f4d2662.png)

**_the plot drawn from t.results_**
![p_orig](https://user-images.githubusercontent.com/14822750/171618174-be9fbf95-9fef-4472-929f-78829b835f35.png)


In the proposed change, we use a preemptive stop of the calculations to keep the temperature of the device within comfortable limits. As a result, the temperature does not exceed 70 degrees.

```
from tvm.contrib.debugger import debug_runtime as graph_executor
...
m = graph_executor.create(graph, lib, ctx)
...
time_f = m.module.time_evaluator("run", dev, number=1, repeat=200, 
                                                       run_delay=1000, repeats_to_cooldown=12)
t =  time_f()
mean = t.mean
0.0832038 secs/iteration

print(t.results)
[0.083007232, ... , 0.083067136, ... , 0.083169024]
```

**_the results of device profiling_** 
![66_patch](https://user-images.githubusercontent.com/14822750/171618146-1a3ea6fc-2e83-48a8-b695-6c4d24ced337.png)

**_the plot drawn from t.results_**
![p_patch](https://user-images.githubusercontent.com/14822750/171618196-732ab833-818e-40c9-bbe4-9d40aa333972.png)


```
from tvm.contrib.debugger import debug_runtime as graph_executor
...
m = graph_executor.create(graph, lib, ctx)
...
m.run(number=100, repeat=2) 
```
   

> Node Name  Ops  Time(us)   Time(%)  Shape   Inputs  Outputs  Measurements                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
> ---------  ---  --------   -------  -----   ------  -------  -------                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
> f1         f1   9451.75    9.822    (2, 2)  3       1        [9.456, 9.465, ..., 9.463]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
> f2         f2   5255.25    5.461    (2, 2)  4       1        [5.261, 5.308, ..., 5.302]  
> 										...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                
> fn         fn   5214.75    5.419    (2, 2)  4       1        [5.266, 5.192, ..., 5.199]
               

**Changes:**

- 	Added `cooldown_interval_ms` and `repeats_to_cooldown` parameter to next functions:
 			_python\tvm\runtime\vm.py:benchmark
 			python\tvm\runtime\module.py:time_evaluator
 			python\tvm\contrib\graph_executor.py:benchmark
			python\tvm\contrib\debugger\debug_executor.py:run
 			python\tvm\contrib\debugger\debug_executor.py:run_individual
 			python\tvm\contrib\debugger\debug_executor.py:run_individual_node_
 		The parameter is used to stop measuring/cooling of the device.
- 	For _python\tvm\contrib\debugger\debug_executor.py:run_ `number`, `repeat`, `min_repeat_ms`, `repeats_to_cooldown` parameters were added.
- 	Fixed _RunIndividual_ logging.
- 	Removed _USE_PROFILER_ that is no longer in use.

### CODE REVIEW

**Comment by Icemist** [comment]
@areusch @tmoreau89 @AndrewZhaoLuo

**Comment by areusch** [comment]
@Icemist i like the direction here--but, i am wondering if it's possible to put some teeth behind this by allowing an application-specific way to verify the cooldown actually happened. for example, perhaps one could pass a PackedFunc that knows how to check the die temperature and continue cooldown if it is still too hot. what do you think of this? otherwise, it strikes me as a feature that would be hard to use in practice if you weren't already watching this separately.

**Comment by Icemist** [comment]
> @Icemist i like the direction here--but, i am wondering if it's possible to put some teeth behind this by allowing an application-specific way to verify the cooldown actually happened. for example, perhaps one could pass a PackedFunc that knows how to check the die temperature and continue cooldown if it is still too hot. what do you think of this? otherwise, it strikes me as a feature that would be hard to use in practice if you weren't already watching this separately.

This cooldown_interval logic already used for RPCRunner, LocalRunner and etc.
The other part of the patch only allows you to verify the quality of the data.

In your suggestion, in theory, you could try to use a ready-made f_preproc. Do 1 repeat, call f_preproc with some code to check and cool the temperature, do the next repeat.
```
for(i : repeat)
 call(f_preproc) // sleep here until the temperature is right  
 timer.start()
 call(pf)
 timer.end()
```
It seems to me that both options have a place.

**Comment by AndrewZhaoLuo** [comment]
cc @tkonolige

File: src/runtime/profiling.cc
Comment (tkonolige) [inline src/runtime/profiling.cc:839]: The reason we have the timer outside the loop is so that we can accurately time very small duration functions. We can amortize the inaccuracy in the timer by running the small duration programs many times under a single timer and then dividing the total time by the number of runs.

File: src/runtime/profiling.cc
Comment (tkonolige) [inline src/runtime/profiling.cc:856]: Can you guard this which a check that `cooldown_iterval_ms` is greater than 0.

**Review by tkonolige — DISMISSED** [review]
@Icemist thank you for the PR. Having a cooldown interval during timing is very useful.

However, you've moved the timer in `time_evaluator` into the inner loop. The timer is specifically in the outer loop so that we can time functions with a very small runtime or use less precise timers. I'll need you to revert this change before we can accept this PR.

**Review by tkonolige — COMMENTED** [review]
Changing my review to "Comment" so it doesn't block the PR while I am out of town for the next couple of weeks. Moving the timers into the inner loop still needs to be reverted though.

**Comment by areusch** [comment]
@Icemist it'd be great if you could address @tkonolige 's comments. 

> In your suggestion, in theory, you could try to use a ready-made f_preproc. Do 1 repeat, call f_preproc with some code to check and cool the temperature, do the next repeat.

just wondering, would it be possible to achieve same effect as this PR with that strategy?

File: python/tvm/contrib/debugger/debug_result.py
Comment (Icemist) [reply python/tvm/contrib/debugger/debug_result.py]: I added a return of all repeats, not just the first. Renamed it "Measurements(us)".

File: src/runtime/graph_executor/debug/graph_executor_debug.cc
Comment (Icemist) [reply src/runtime/graph_executor/debug/graph_executor_debug.cc:102]: Done

File: src/runtime/graph_executor/debug/graph_executor_debug.cc
Comment (Icemist) [reply src/runtime/graph_executor/debug/graph_executor_debug.cc]: Done

File: src/runtime/graph_executor/debug/graph_executor_debug.cc
Comment (Icemist) [reply src/runtime/graph_executor/debug/graph_executor_debug.cc]: Done

File: src/runtime/profiling.cc
Comment (Icemist) [reply src/runtime/profiling.cc]: Reverted

File: src/runtime/profiling.cc
Comment (Icemist) [reply src/runtime/profiling.cc]: Done

File: src/runtime/crt/common/crt_runtime_api.c
Comment (Icemist) [reply src/runtime/crt/common/crt_runtime_api.c]: Fixed

File: python/tvm/contrib/debugger/debug_result.py
Comment (Icemist) [reply python/tvm/contrib/debugger/debug_result.py]: Before my changes, it was an average of 1 repeat. Here. I returned not only the first repeat of the node, but all of them for each node, and also in the form of arrays as raw measurements.
Now I just return all repeats for each node, not just the first. So I take the average here. Below where the duration is calculated, I also calculate the average.

**Comment by Icemist** [comment]
> @Icemist it'd be great if you could address @tkonolige 's comments.
> 
> > In your suggestion, in theory, you could try to use a ready-made f_preproc. Do 1 repeat, call f_preproc with some code to check and cool the temperature, do the next repeat.
> 
> just wondering, would it be possible to achieve same effect as this PR with that strategy?

It seems to me with that strategy it is more difficult to get a quick and generalized solution.
There may be a need for some general mechanics of adding such functions for different devices by developers familiar with these systems.

**Comment by Icemist** [comment]
@areusch and  @AndrewZhaoLuo, what do you think? I addressed all the remarks. I also followed up on @tkonolige's request.

File: include/tvm/runtime/profiling.h
Comment (Icemist) [reply include/tvm/runtime/profiling.h:561]: done

File: src/runtime/crt/common/crt_runtime_api.c
Comment (Icemist) [reply src/runtime/crt/common/crt_runtime_api.c:564]: moved it up to the right place

File: src/runtime/graph_executor/debug/graph_executor_debug.cc
Comment (Icemist) [reply src/runtime/graph_executor/debug/graph_executor_debug.cc:64]: I wonder what caused it. I didn't have any problems even when there were tens of thousands of numbers here.
I would guess that the right thing to do here would be to pack the data as it is done in `time_evaluator`. And then extract the data with `struct.unpack`.
I would like to know the reason for the comment:
// Have problems returning FloatImm so serialize to string results as hack.
@AndrewZhaoLuo can you help with that?

In my previous version of the patch, I did dimanic array packing, and this was not a problem for `time_evaluator`.

File: include/tvm/runtime/profiling.h
Comment (Icemist) [reply include/tvm/runtime/profiling.h:579]: done

File: src/runtime/crt/common/crt_runtime_api.c
Comment (tkonolige) [inline src/runtime/crt/common/crt_runtime_api.c:563]: I know the previous code said `1.618` was random, but its actually not :). Its the golden ratio!

File: src/runtime/graph_executor/debug/graph_executor_debug.cc
Comment (AndrewZhaoLuo) [inline src/runtime/graph_executor/debug/graph_executor_debug.cc:98]: Probably want full precision so want something like

```
// use maximum precision available and use fixed representation
s << std::fixed;
s.precision(std::numeric_limits<double>::max_digits10);
```

and feed every double manually into the stream.

File: src/runtime/graph_executor/debug/graph_executor_debug.cc
Comment (AndrewZhaoLuo) [reply src/runtime/graph_executor/debug/graph_executor_debug.cc:64]: I forget the exact reason why FloatImm can't be serialized but what you are doing is the string serialization hack which is fine. 

"returning a long string over rpc can be expensive." i don't have context on but if you aren't having any problems it's probably fine?

**Comment by AndrewZhaoLuo** [comment]
Leave the rest up to @tkonolige

**Comment by Icemist** [comment]
@tkonolige I added a commit to this PR https://github.com/apache/tvm/pull/11465/commits/3e3b2cdd242089f777d969893e22e9b4d8fab517 where I change the way the RunIndividual and RunIndividualNode functions get results. I think this should solve the issue above.

File: src/runtime/graph_executor/debug/graph_executor_debug.cc
Comment (tkonolige) [inline src/runtime/graph_executor/debug/graph_executor_debug.cc:202]:
```suggestion
    LOG(INFO) << "Got op timing: " << os.str();
```

**Review by tkonolige — APPROVED** [review]
Looks great! Thanks @Icemist!

**Pull request merged**
